vcr/oidc4vci/issuer_client.go

```diff
@@ -0,0 +1,211 @@
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ */
+
+package oidc4vci
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"github.com/nuts-foundation/go-did/vc"
+	"github.com/nuts-foundation/nuts-node/vcr/log"
+	"io"
+	"net/http"
+	"net/url"
+	"strings"
+)
+
+// IssuerAPIClient defines the API client used by the wallet to communicate with the credential issuer.
+type IssuerAPIClient interface {
+	OAuth2Client
+
+	Metadata() CredentialIssuerMetadata
+	RequestCredential(ctx context.Context, request CredentialRequest, accessToken string) (*vc.VerifiableCredential, error)
+}
+
+// NewIssuerAPIClient resolves the Credential Issuer Metadata from the well-known endpoint
+// and returns a client that can be used to communicate with the issuer.
+func NewIssuerAPIClient(ctx context.Context, httpClient *http.Client, credentialIssuerIdentifier string) (IssuerAPIClient, error) {
+	if credentialIssuerIdentifier == "" {
+		return nil, errors.New("empty Credential Issuer Identifier")
+	}
+
+	// Load OIDC4VCI metadata and OIDC metadata
+	// TODO: Use the OIDC4VCI credential issuers metadata to load the OIDC metadata?
+	metadata, err := loadCredentialIssuerMetadata(ctx, credentialIssuerIdentifier, httpClient)
+	if err != nil {
+		return nil, fmt.Errorf("unable to load Credential Issuer Metadata (identifier=%s): %w", credentialIssuerIdentifier, err)
+	}
+	providerMetadata, err := loadOIDCProviderMetadata(ctx, credentialIssuerIdentifier, httpClient)
+	if err != nil {
+		return nil, fmt.Errorf("unable to load OIDC Provider Metadata (identifier=%s): %w", credentialIssuerIdentifier, err)
+	}
+	return NewIssuerClientFromMD(httpClient, *providerMetadata, *metadata)
+}
+
+// NewIssuerClientFromMD creates a new IssuerAPIClient from preloaded metadata.
+func NewIssuerClientFromMD(httpClient *http.Client, oidcProvider ProviderMetadata, credentialIssuer CredentialIssuerMetadata) (IssuerAPIClient, error) {
+	return &defaultIssuerAPIClient{
+		httpOAuth2Client: httpOAuth2Client{
+			httpClient: httpClient,
+			metadata:   oidcProvider,
+		},
+		identifier: credentialIssuer.CredentialIssuer,
+		httpClient: httpClient,
+		metadata:   credentialIssuer,
+	}, nil
+}
+
+var _ IssuerAPIClient = (*defaultIssuerAPIClient)(nil)
+
+type defaultIssuerAPIClient struct {
+	httpOAuth2Client
+
+	identifier       string
+	httpClient       *http.Client
+	metadata         CredentialIssuerMetadata
+	providerMetadata ProviderMetadata
+}
+
+func (h defaultIssuerAPIClient) RequestCredential(ctx context.Context, request CredentialRequest, accessToken string) (*vc.VerifiableCredential, error) {
```

Comment:
A generated client assumes there's a complete generic OpenAPI spec of OpenID4VCI, while OpenAPI isn't a very good fit for protocol with very dynamic URLs (since everything depends on the URLs in the metadata). We could still use the gnerated models, but those are already hand-crafted. What a generated client would bring is then so thin of a layer, then I prefer a hand-crafted, isolated and well-tested client. This package is set up such that it could be a standalone client library(/repo) on its own. No strings attached.